receiver/snmpreceiver/scraper.go

```diff
@@ -39,13 +90,489 @@ func newScraper(logger *zap.Logger, cfg *Config, settings component.ReceiverCrea
 }
 
 // start gets the client ready
-func (s *snmpScraper) start(_ context.Context, _ component.Host) (err error) {
-	return nil
+func (s *snmpScraper) start(_ context.Context, host component.Host) (err error) {
+	s.client, err = newClient(s.cfg, s.logger)
+
+	return err
 }
 
 // scrape collects and creates OTEL metrics from a SNMP environment
 func (s *snmpScraper) scrape(_ context.Context) (pmetric.Metrics, error) {
+	if err := s.client.Connect(); err != nil {
+		return pmetric.NewMetrics(), fmt.Errorf("problem connecting to SNMP host: %w", err)
+	}
+	defer s.client.Close()
+
+	// Get a basic ResourceMetrics prepped for metrics with no resource attributes
+	resourceMetrics := pmetric.NewResourceMetrics()
+	scopeMetrics := resourceMetrics.ScopeMetrics().AppendEmpty()
+	scopeMetrics.Scope().SetName("otelcol/snmpreceiver")
+	scopeMetrics.Scope().SetVersion(s.settings.BuildInfo.Version)
+	metricSlice := scopeMetrics.Metrics()
+	now := pcommon.NewTimestampFromTime(time.Now())
+
+	resourceMetricsMap := map[string]*pmetric.ResourceMetrics{}
+	metricsMap := map[string]*pmetric.Metric{}
+	// Try to scrape scalar OID based metrics.
+	// metricsMap is passed in as a place for any created metrics to live
+	var scraperErrors scrapererror.ScrapeErrors
+	s.scrapeScalarMetrics(now, metricsMap, &scraperErrors)
+	if len(metricsMap) != 0 {
+		// Add new scalar metrics to the basic ResourceMetrics with no resource attributes
+		for _, value := range metricsMap {
+			value.MoveTo(metricSlice.AppendEmpty())
+		}
+		// Load the map with the basic ResourceMetrics (that possibly
+		// has scalar OID based metrics already on it)
+		resourceMetricsMap[""] = &resourceMetrics
+	}
+
+	// Load this with the scalar metric map that was just created for the
+	// basic ResourceMetrics. The info in this map is ultimately already
+	// contained within the resourceMetricsMap, but it is more easily accessible
+	// to pull out a specific existing Metric by resource and metric name using this
+	metricsByResource := newMetricsByResourceContainer()
+	metricsByResource.putMetrics(generalResourceKey, metricsMap)
+
+	// Try to scrape column OID based metrics.
+	// resourceMetricsMap is passed in as a place for any created resources/metrics to live
+	// metricsByResource is provided as an easy way to check if a metric is already
+	// associated with an existing resource
+	s.scrapeIndexedMetrics(now, resourceMetricsMap, metricsByResource, &scraperErrors)
+
+	// Return error if we failed to scrape any metrics
 	md := pmetric.NewMetrics()
 
-	return md, nil
+	// Put all of the created ResourceMetrics onto a top level Metrics
+	for _, value := range resourceMetricsMap {
+		value.MoveTo(md.ResourceMetrics().AppendEmpty())
+	}
```

Comment:
@djaglowski i hear you. again there is no way to grab existing ResourceMetric from a ResourceMetricsSlice. and because there is code that is being performed on each piece of SNMP data that was retrieved, i have maps to store detached versions of these structures and then later assign them to the final holder that is actually being returned from the scrape